taf/tests/test_updater/test_update_library/test_validation_and_sync.py

```diff
@@ -0,0 +1,214 @@
+    we should not make this more complicated by test the update of a repo with dependencies).
+    Make valid changes to both the auth repo and the target repos (only origin repos).
+    Sync client's auth repo with remote without running the updater.
+    Run the updater. Expect and check that all repositories were successfully updated to the last remote commit.
+    """
+    clone_repositories(
+        origin_auth_repo,
+        client_dir,
+    )
+    setup_manager = SetupManager(origin_auth_repo)
+    setup_manager.add_task(update_expiration_dates)
+    setup_manager.add_task(add_valid_target_commits)
+    setup_manager.add_task(sync_auth_repo)
+    setup_manager.execute_tasks()
+
+    update_and_check_commit_shas(
+        OperationType.UPDATE,
+        origin_auth_repo,
+        client_dir,
+    )
+
+
+@pytest.mark.parametrize(
+    "origin_auth_repo",
+    [
+        {
+            "targets_config": [
+                {"name": "target1", "allow_unauthenticated_commits": True},
+                {"name": "target2", "allow_unauthenticated_commits": True},
+            ],
+        },
+    ],
+    indirect=True,
+)
+def test_target_repo_not_in_sync(origin_auth_repo, client_dir):
+    """
+    Client's authentication repository is on last validated commits,
+    but target repositories contain commits following the last validated ones.
+    Simulate the case when someone used git pull to update the one or more target repositories.
+    Run the updater to clone all repositories. Make valid changes to both the auth repo and the target repos (only origin repos).
+    Sync one or more target repos with remote without running the updater. Run the updater.
+    Expect and check that all repositories were successfully updated to the last remote commit.
+    """
+    clone_repositories(
+        origin_auth_repo,
+        client_dir,
+    )
+    setup_manager = SetupManager(origin_auth_repo)
+    setup_manager.add_task(add_valid_target_commits)
+    setup_manager.add_task(update_expiration_dates)
+    setup_manager.add_task(
+        sync_target_repos_with_remote,
+        kwargs={"origin_auth_repo": origin_auth_repo, "client_dir": client_dir},
+    )
+    setup_manager.execute_tasks()
+
+    # Run the updater to update repositories
+    update_and_check_commit_shas(
+        OperationType.UPDATE,
+        origin_auth_repo,
+        client_dir,
+    )
+
+
+@pytest.mark.parametrize(
+    "origin_auth_repo",
+    [
+        {
+            "targets_config": [
+                {"name": "target1", "allow_unauthenticated_commits": True},
+                {"name": "target2", "allow_unauthenticated_commits": True},
+            ],
+        },
+    ],
+    indirect=True,
+)
+def test_auth_repo_not_in_sync_partial(origin_auth_repo, client_dir):
+    """
+    Clients auth repo has been manually updated, just like in 1.
+    The update should result in a partial update.
+    Clients auth repo should be reverted to the last valid commit,
```

Comment:
Sorry, I not explain this too well. What we want to check here is if th updater reverts the local git history to the last commit that was validated. So, add at least one invalid commit and pull manually. The updater should basically remove they invalid commit that was pulled